homeassistant/components/kodi/kodi_connman.py

```diff
@@ -0,0 +1,325 @@
+
+    async def _async_ws_connect(self):
+        """Connect to Kodi via websocket protocol."""
+        try:
+            await self._connection.connect()
+            await self._on_ws_connected()
+        except (TransportError, CannotConnectError):
+            if not self._connect_error:
+                _LOGGER.warning("Unable to connect to Kodi via websocket")
+            await self._on_ws_error(False)
+        else:
+            self._connect_error = False
+
+    async def _ping(self) -> bool:
+        """Check connectivity using Kodis buildin ping method."""
+        try:
+            await self.kodi.ping()
+        except (TransportError, CannotConnectError):
+            return False
+        else:
+            return True
+
+    async def _async_connect_websocket_if_disconnected(self, *_):
+        """Reconnect the websocket if it fails."""
+        if not self.connected:
+            await self._async_ws_connect()
+        elif not await self._ping():
+            if not self._connect_error:
+                _LOGGER.warning("Unable to ping Kodi via websocket")
+            await self._on_ws_error()
+        else:
+            self._connect_error = False
+
+    async def _ws_callback_dummy(self, sender: str, data: str) -> None:
+        """Use to 'unregister' websocket callbacks.
+
+        Since jsonrpc_base doesn't offer an unregister method, an
+        empty function is used as a replacement.
+        """
+
+    def register_websocket_callback(
+        self, api_method: str, func: Callable[[str, str], Awaitable[None]]
+    ) -> None:
+        """Register a callback for a websocket server request."""
+        setattr(self._connection.server, api_method, func)
+
+    def unregister_websocket_callback(self, api_method: str) -> None:
+        """Unregister a callback for a websocket server request."""
+        setattr(self._connection.server, api_method, self._ws_callback_dummy)
+
+        # del self._connection.server._server_request_handlers[method_name]
+        # would also work, but this access to a protected member is bad practice
+        # and also gives a lint error.
+
+    async def add_callback_on_connect(
+        self, func: Callable[..., Awaitable[None]]
+    ) -> None:
+        """Add a function to call when ws is connected."""
+        self._cb_on_ws_connect.append(func)
+
+        if not self._connection.can_subscribe:
+            return
+        # Call callback when added since ws might be already connected
+        if self._connection.connected:
+            await func()
+
+    async def add_callback_on_disconnect(
+        self, func: Callable[..., Awaitable[None]]
+    ) -> None:
+        """Add a function to call when ws is disconnected."""
+        self._cb_on_ws_disconnect.append(func)
+
+
+class KodiConnectionClient(Entity):
+    """Base class to be used by Kodi platforms."""
+
+    _attr_has_entity_name = True
+    _websocket_callbacks: dict[str, Callable[[Any, Any], Awaitable[None]]] = {}
```

Comment:
Shouldn't that be at the instance level, rather than the class?

Reply:
Probably also makes sense accepting these in `__initi__()`

Reply:
I took an example from the  Entity class variable [_on_remove](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py#L264) and thought extending the Entity with my list of callbacks, mimicking the same behaviour. But setting it directly might not be the best choice. I will add the dict to the `__init__()` method as parameter.